### PR TITLE
Jesus Meeting Query - No Optional Attendees

### PR DESCRIPTION
**Pull request by jelares**

### CODE REVIEW

File: jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -15,9 +15,149 @@
...
+
+        // Case 3: if current timeRange is contained within current condensed time range being built, do nothing
+      }
+    }
+
+    return condensedTimeRanges;
+  }
+
+  /**
+   * Finds the timeRanges in the day which are not covered by any events. May add ranges of duration 0
```
Comment (ykjchen) [inline jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:115]: Perhaps you could clarify `May add ranges of duration 0 between timeRanges.`

Do you just mean that some ranges may have a duration of 0?

File: jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -15,9 +15,149 @@
...
+
+    // Sort the timeRanges by start time.
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+    return timeRanges;
+  }
+
+  /**
+   * Condenses an ordered list of time ranges (ordered by start time).
+   * @param timeRanges list of timeRanges sorted by start time
+   * @return list of condensed timeRanges (condensed means overlapping events are turned into a single event).
```
Comment (ykjchen) [inline jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:80]: Are adjacent events considered overlapping? e.g. would an 8:00->8:30 event be condensed with an 8:30->9:00 event? May be worth pointing this out in the docs.

File: jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,149 @@
...
+    // sorted by their start time. Inverse means all the time ranges not covered by any time range
+    // in the collection.
+    List<TimeRange> openTimeRanges = getInverseTimeRanges(condensedTimeRanges);
+
+    // Filter the open time ranges such that only open time ranges with a duration greater than the duration of the
+    // meeting remain
+    List<TimeRange> validTimeRanges = getValidTimeRanges(openTimeRanges, (int) request.getDuration());
+
+    return validTimeRanges;
   }
```
[the remark this answers is not in the document]
Comment (jelares) [reply jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:47]: Thanks!

File: jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,149 @@
...
+          // Add the timeRange as a new condensed range
+          condensedTimeRanges.add(timeRange);
+        }
+        // Case 2: current timeRange overlaps with current condensed time range being built, but is not contained in it
+        else if (currentCondensedRange.overlaps(timeRange) && !currentCondensedRange.contains(timeRange)) {
+          // Replace the current condensed time range with a new condensed time range augmented with the current timeRange
+          TimeRange newCondensedRange = TimeRange.fromStartEnd(currentCondensedRange.start(), timeRange.end(), false);
+
+          condensedTimeRanges.remove(condensedTimeRanges.size()-1);
+          condensedTimeRanges.add(newCondensedRange);
```
[the remark this answers is not in the document]
Comment (jelares) [reply jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:104]: I have an idea, which I left as a TO-DO

File: jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (jelares) [reply jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:80]: No - adjacent events are not condensed. This follows the definition of "overlapping timeRanges" from the timeRange class. I updated the documentation to include this information.

File: jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (jelares) [reply jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:115]: I clarified this point: an inverse range of duration zero will be added to represent the space between adjacent events.

File: jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,149 @@
...
+    Collections.sort(timeRanges, TimeRange.ORDER_BY_START);
+    return timeRanges;
+  }
+
+  /**
+   * Condenses an ordered list of time ranges (ordered by start time).
+   * @param timeRanges list of timeRanges sorted by start time
+   * @return list of condensed timeRanges (condensed means overlapping events are turned into a single event).
+   */
+  private List<TimeRange> getCondensedTimeRanges(List<TimeRange> timeRanges) {
```
[the remark this answers is not in the document]
Comment (jelares) [reply jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:82]: I updated all the function names to match this convention.

File: jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -15,9 +15,149 @@
...
+    List<TimeRange> openTimeRanges = getInverseTimeRanges(condensedTimeRanges);
+
+    // Filter the open time ranges such that only open time ranges with a duration greater than the duration of the
+    // meeting remain
+    List<TimeRange> validTimeRanges = getValidTimeRanges(openTimeRanges, (int) request.getDuration());
+
+    return validTimeRanges;
   }
+
+  // Adds all relevant events (events with at least one required attendee) to a new collection.
```
[the remark this answers is not in the document]
Comment (jelares) [reply jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:49]: Sounds good - I updated all the JavaDocs to follow this convention.

File: jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -15,9 +15,149 @@
...
+  private Collection<Event> getRelevantEvents(Collection<Event> events, Collection<String> requiredAttendees) {
```
[the remark this answers is not in the document]
Comment (jelares) [reply jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: I updated this method name to more properly reflect the action which is taking place.

File: jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ykjchen) [reply jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:106]: I see... another potential optimization may be to introduce a local variable on line 86 that stores the current "in progress" time range. You could just continue to reassign that local variable to `newCondensedRange`, and add it to `condensedTimeRanges` only once you've detected a lack of overlap. This I think would avoid the repeated remove+add operations.

[the remark this answers is not in the document]
Comment (jelares) [reply jelares/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:106]: Nice!! That's definitely the most elegant solution

**Pull request merged**